cpp/src/arrow/acero/assert_order_node.cc

```diff
@@ -0,0 +1,402 @@
+
+  virtual ~ColumnComparator() = default;
+
+  virtual int Compare(const Array& column, const int64_t& left,
+                      const int64_t& right) const = 0;
+
+  compute::NullPlacement null_placement;
+  compute::SortOrder sort_order;
+  int left_is_null_order;
+  int right_is_null_order;
+};
+
+template <typename Type>
+struct ConcreteColumnComparator final : public ColumnComparator {
+  using ColumnComparator::ColumnComparator;
+
+  using ArrayType = typename TypeTraits<Type>::ArrayType;
+  using GetView = compute::internal::GetViewType<Type>;
+
+  int Compare(const Array& column, const int64_t& left,
+              const int64_t& right) const override {
+    const ArrayType& array = ::arrow::internal::checked_cast<const ArrayType&>(column);
+
+    const auto left_value = GetView::LogicalValue(array.GetView(left));
+    const auto right_value = GetView::LogicalValue(array.GetView(right));
+
+    auto left_is_null = column.IsNull(left);
+    auto right_is_null = column.IsNull(right);
+
+    auto orientation = sort_order == compute::SortOrder::Ascending ? 1 : -1;
+    auto less = -1 * orientation;
+    auto larger = 1 * orientation;
+
+    if (left_is_null && right_is_null) {
+      // both values are null (equal)
+      return 0;
+    } else if (left_is_null) {
+      // left value is null
+      return left_is_null_order;
+    } else if (right_is_null) {
+      // right value is null
+      return right_is_null_order;
+    } else if (left_value < right_value) {
+      return less;
+    } else if (left_value > right_value) {
+      return larger;
+    }
+
+    return 0;
+  }
+};
+
+template <>
+struct ConcreteColumnComparator<NullType> : public ColumnComparator {
+  using ColumnComparator::ColumnComparator;
+
+  int Compare(const Array& column, const int64_t& left,
+              const int64_t& right) const override {
+    return 0;
+  }
+};
+
+class AssertOrderNode : public ExecNode,
+                        public TracedNode,
+                        util::SequencingQueue::Processor {
+ public:
+  AssertOrderNode(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                  std::shared_ptr<Schema> output_schema, const Ordering& ordering)
+      : ExecNode(plan, std::move(inputs), {"input"}, std::move(output_schema)),
+        TracedNode(this),
+        ordering_(ordering),
+        comparators_(MakeComparators(output_schema_, ordering_)),
+        sort_ascs_(MakeSortAscs(ordering.sort_keys())),
+        sequencing_queue_(util::SequencingQueue::Make(this)) {}
+
+  static Result<ExecNode*> Make(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                                const ExecNodeOptions& options) {
+    RETURN_NOT_OK(ValidateExecNodeInputs(plan, inputs, 1, "AssertOrderNode"));
+
+    const auto& assert_options = checked_cast<const AssertOrderNodeOptions&>(options);
+
+    Ordering ordering = assert_options.ordering;
+
+    // check output ordering
+    if (ordering.is_implicit() || ordering.is_unordered()) {
```

Comment:
Should you use `!ordering.is_explicit()`?

Reply:
right, fixed